### DISCUSSION TITLE
ConnectionRefused when running Jupyter Lab and tutorial

### DISCUSSION THREAD

**Discussion by LeighTilley** [atoti/atoti#882] [Q&A]
Hey

I hope you are all well! :)

I had some time to return to the Python Atoti tutorial. I just forced an update to get 0.8.12. If I run Jupyter Lab and then the start of the tutorial to create, I see:

![image](https://github.com/atoti/atoti/assets/9533038/0bc8de37-df8b-4505-b7ef-023d4a0bdc40)

Before I dig further, is there anything else I need to do/update?

Thanks

Leigh
https://tilleytech.com

**Comment by LeighTilley** [comment]
Ah

I restarted my PC and I am seeing this popup when I run Jupyter Lab 

![image](https://github.com/atoti/atoti/assets/9533038/c63def6a-c53f-49f1-858a-90343dfd0e42)

So do I need to clear some cache or something?

Thanks

Leigh
https://tilleytech.com

**Comment by LeighTilley** [comment]
Hi

I just tried --port 8889 instead, and I still see the above errors, firstly 'Plugin 'atoti' is already registered.' and then the ConnectioNRefused error so I am unable to use it.

![image](https://github.com/atoti/atoti/assets/9533038/f9a1da46-480e-4047-8436-787fa966a0e1)


Please let me know if it's a known issue or it's something else Python related I need to do (and I don't use Python much yet :)).

Thanks

Leigh
https://tilleytech.com

**Comment by tibdex** [comment]
I can't reproduce your issue on a fresh install of Atoti. I ran `poetry install atoti[jupyterlab]` in a new virtual environment with Python 3.9.18 and Poetry 1.8.2 on macOS. Can you try the same command and if it still doesn't work give us details about your environment?

**Reply by LeighTilley** [reply]
Hey

OK I'm on Python 3.10.11 and I use pip normally. Should I get poetry then?

**Reply by tibdex** [reply]
Yes, to debug this it will be easier with Poetry since it guarantees (by default) that each project uses a different virtual environment.

**Reply by LeighTilley** [reply]
Hey

OK. So I got poetry, and then I tried your above and got a msg re it could not find a project.toml file. I looked this up and found I need to do init.

It asked me if I want to interactively select packages and it listed some atoti ones. Do I need to add them all?

I went back to your original Atoti install page and did this - poetry add "atoti[jupyterlab]" and this downloaded much more stuff (although it did error on installing Jedi).

How do I know / start a fresh run though? (Sorry if this is basic Python questioning :)). As I only know how to run 'jupyter lab' and I guess I need to clear everything.

**Reply by LeighTilley** [reply]
OK 

I looked stuff up and did:

poetry add jupyter
poetry add jupyterlab

I only see errors regarding jedi so will look that up.

If I now do:

poetry run jupyter lab

I still see 'atoti plugin already registered. And I get the same ConnectionRefused error too.

**Reply by LeighTilley** [reply]
Hey

And here is cmd I just discovered, poetry env info:

![image](https://github.com/atoti/atoti/assets/9533038/490145d8-ab21-41b6-9401-ceb4365c4a4a)

**Comment by andrewyang94** [comment]
Hi @LeighTilley, thanks for posting about your issue. Let's just take a step back real quick, can you execute the following commands for me? For context, we're going to do a fresh poetry install of the dependencies listed in https://github.com/atoti/atoti. Please see https://python-poetry.org/docs/#installation if you need help installing `poetry`, but it seems like you already have it installed on your machine.

```
git clone https://github.com/atoti/atoti.git
```

Assuming you've already cloned our notebook gallery repo and run a `poetry install` in the past, rename the newly cloned `atoti` folder on your machine to `atoti-debug`. This ensures that when we run the following `poetry install` at the root of the (renamed) repo folder, we're creating a new poetry environment. If you are cloning our repo for the first time, you can just keep the name of the folder as-is.

Now run the rest of the commands. In particular, we'll be doing a change directory into the folder (at the root of the repo), running `poetry install` to install the dependencies, and then running `jupyter-lab` in the poetry-managed environment.

```
cd atoti-debug
poetry install
```

This should now have created a new Poetry-managed environment for you, and you can check the env info and python dependencies with the following commands:

```
poetry env info
poetry run pip list | grep atoti <-- this may be `findstr` in windows
```

Finally, let's try the tutorial again with the correct dependencies installed.

```
poetry run python -m atoti.copy_tutorial tutorial
cd tutorial
poetry run jupyter-lab
```

If you're continuing to have issues, can you post the full log of the error stacktrace? It seems the first `ConnectionRefused` error had some more log content that I couldn't see from your screenshot.

Thanks,
Andrew

**Reply by LeighTilley** [reply]
Hey

OK thanks will do now.

I didn't actually have git on this machine, so I downloaded that and then cloned atoti.

![image](https://github.com/atoti/atoti/assets/9533038/e38d3adc-943b-4d7a-872f-298ebcf09c47)

And then poetry install:

![image](https://github.com/atoti/atoti/assets/9533038/d2e6befe-6568-40eb-8211-6cabff51cecf)

![image](https://github.com/atoti/atoti/assets/9533038/2a0c59fa-3b27-4b3b-b9c3-bfe6f2913405)

Env info shows this, and listing atoti dependencies does not find anything:

![image](https://github.com/atoti/atoti/assets/9533038/a703813a-6c7f-4fdd-b7e3-adb0791ea8c9)

**Comment by LeighTilley** [comment]
Hi

I am guessing you are busy :); no worries.

In meantime just now I put it all on my laptop, as the laptop has none of this tech on it yet (python, pip, poetry, atoti etc). And it's working as expected with all libs at 0.8.12.

So I will blitz this big main PC as prob a better use of time to blitz it, clean and restart rather than stay stuck in some sort of corrupt setup! :)

**Reply by andrewyang94** [reply]
Ah apologies, I had forgotten to reply yesterday! In terms of your previous failing setup, I think the issue was that for whatever reason your machine was unable to install all the dependencies necessary to run the tutorials properly (looks like `jedi` was having an issue installing in both cases).

For clarity, I was going to ask what version of `python`, `pip`, `poetry`, and Windows OS you were running so that I could attempt to re-create your issue, but if you've managed to unblock yourself by installing everything on a separate machine, thats okay too ☺️ 

Thanks again for reaching out! Please let us know if you have any further issues 🙏

**Reply by LeighTilley** [reply]
Yeah no worries mate.

Thx for prompt reply and may meet you if you go to the London Connect in June.

Thanks

Leigh
https://tilleytech.com

**Comment by LeighTilley** [comment]
Hey @andrewyang94 

Solved - in a previous post / thread on here I had an issue with a previous release and Thibault had directed me through it. Related to use of atoti-jupyterlab3 0.8.3,  and as this was present when atoti-jupyterlab 0.8.12 came down they clashed!

**Reply by andrewyang94** [reply]
😮 That would make sense! Thanks for chiming back in and updating us, we'll be more mindful of things like this in the future when similar situations arise 🙏 But glad you figured it out! 🍻

**Reply by LeighTilley** [reply]
I still need to get rest of big PC here working, and the above solves the 'atoti plugin already registered' issue.

For your poetry guide above,which worked fine on laptop, does not work here due to Jedi. So just ned to work that out. I think almost there on this PC though.

**Reply by andrewyang94** [reply]
Sounds good. Yeah it's interesting, I'm wondering what are the differences between your laptop and your big PC? Perhaps it just comes down to different versions of `python`, `pip`, `poetry`, and `Windows OS`? 🤔

**Reply by LeighTilley** [reply]
Hey mate

OK so back on big PC, had a bit of time to check it. I checked the error stack trace more in depth and much lower down past the 'ConnectionRefused' you actually output where you go looking for Java. I remembered I've not done any Java on here for a very long time so I betted it'd be out of date..yeah 8 haha. I saw this by running the StoryTelling book as I hadn't seen them before.

Grabbed AdoptOpenJDK / Eclipse Adoptium 11 and now error has moved to 'missing plugins' which is expected as the Poetry install failed due to Jedi...altho weirdly if I do pip freeze | findstr it has the version required...so will try to fix that.

Or actually checking the laptop, that also had 8 on, but no JAVA_HOME set, so perhaps if you do not find JAVA_HOME you default to something else. Anyway, 11 on both this big PC and laptop now. On laptop Atoti tutorial working perfectly as before, on big PC I just need to sort out this Jedi weirdness and get the core atoti components on.

**Reply by LeighTilley** [reply]
Hey @andrewyang94 

OK so as I do not like to leave stuff I just sorted it all out. :)

I reckon python 3.10.x was a bit weird on here. Laptop has 3.11.9 so I blitzed python, got 3.11.9, got poetry again and bingo; all good. Full install.

I just copied tutorial, ran it and then connected Excel to it too.

![image](https://github.com/atoti/atoti/assets/9533038/0fb81fa5-64ec-486f-b00f-94e8348167a1)

**Reply by andrewyang94** [reply]
Amazing!! Glad to hear everything worked out! And thanks for chiming in, this is all valuable feedback. Cheers! 🍻
